mantidproject/mantid
# Use Jemalloc on Linux

**Description of work.**
Set Jemalloc as the new malloc implementation for Linux, instead of Tcmalloc, so the unused memory is released to the system during run time. See #29054 for details on why this is required.

This adds jemalloc to the list of required packages, and ensures it is used as the memory allocator via CMake.

Note: after merging this, jemalloc will be necessary in order to build master.



**To test:**
To check package builds : 
- Clear CMake configuration and run CMake with flags `ENABLE_CPACK` and `USE_JEMALLOC` on.
- Run `cmake --build . --target package`.
- If necessary, install Jemalloc as `libjemalloc-dev` on Ubuntu or `jemalloc-devel` on RedHat/CentOs.
- Unpackage the created package.

To check Jemalloc works : 
- Open workbench
- Run 

<details>
<summary>memory_report.py</summary>

```
# mem_report.py
import mantid
from mantid.kernel import MemoryStats
from mantid.api import FrameworkManager
from mantid.simpleapi import Load, CreateSampleWorkspace
import os
from time import sleep

def getStats():
    mem = MemoryStats()
    res1 = mem.residentMem()/1024.
    print("Resident Memory Usage (MB)", res1)
    print()

    filename = "MER06398.raw" # in systemtests/Data
    ws = Load(Filename=filename)
    mem.update()
    res2 = mem.residentMem()/1024.
    print("Resident Memory Usage (MB)",res2)
    print("Memory used (MB):",(res2-res1))
    print()

    print("Clearing all memory (workspaces etc)")
    ws.delete()
    FrameworkManager.clear()
    sleep(10)
    mem.update()
    res3 = mem.residentMem()/1024.
    print("Resident Memory Usage (MB)",res3)
    print("Memory change (MB)",(res3-res2))
    print()
    print("Loading file again...")
    ws = Load(Filename=filename)
    mem.update()
    res4 = mem.residentMem()/1024.
    print("Resident Memory Usage (MB)",res4)
    print("Memory change (MB):",(res4-res3))
    print()

    print("Clearing all memory (workspaces etc)")
    FrameworkManager.clear()
    sleep(10)
    mem.update()
    res5 = mem.residentMem()/1024.
    print("Resident Memory Usage (MB)",res5)

    print()
    print("Overall change (MB):",(res5-res1))
    
getStats()
```

</details>

- In the end, the memory should have gone back to its starting value, meaning it has been released to the system.
(The sleeps are necessary because Jemalloc releases the memory progressively, thus if the memory is checked too quickly it is probable the change won't have happened yet.)
 
Fixes #29054 . 


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
